astropy/io/fits/hdu/hdulist.py

```diff
@@ -194,25 +273,30 @@ def __getitem__(self, key):
             hdus = super(HDUList, self).__getitem__(key)
             return HDUList(hdus)
 
-        idx = self.index_of(key)
-        return super(HDUList, self).__getitem__(idx)
+        # Originally this used recursion, but hypothetically an HDU with
+        # a very large number of HDUs could blow the stack, so use a loop
+        # instead
+        return self._try_while_unread_hdus(super(HDUList, self).__getitem__,
```

Comment:
It is odd. In the test that fails (https://github.com/astropy/astropy/blob/master/astropy/io/fits/tests/test_image.py#L1632) you end up with two copies of the file in the HDUList (so `hdus[2]` is another `PrimaryHDU`). Doing almost anything (printing `hdus`, finding its `len` or `repr`) after opening `hdus` (line 1653 of the test)  but before writing it out makes the test pass, but no idea why.